File: 程式設計：資料結構答案檔/C#/205.cs
using System;
using System.Collections.Generic;

public class Program
{
    public class Rectangle
    {
        public int Area, Top, Left, Bottom, Right;

        public Rectangle(int area, int top, int left, int bottom, int right)
        {
            Area = area;
            Top = top;
            Left = left;
            Bottom = bottom;
            Right = right;
        }
    }

    public static Rectangle LargestRectangleArea(List<int> heights, int row)
    {
        int n = heights.Count;
        List<int> left = new List<int>(new int[n]);
        List<int> right = new List<int>(new int[n]);
        for (int i = 0; i < n; i++) right[i] = n;
        Stack<int> stk = new Stack<int>();
        Rectangle maxRect = new Rectangle(0, row, 0, row, 0);

        for (int i = 0; i < n; ++i)
        {
            while (stk.Count > 0 && heights[stk.Peek()] >= heights[i])
            {
                right[stk.Peek()] = i;
                stk.Pop();
            }
            left[i] = stk.Count == 0 ? 0 : stk.Peek() + 1;
            stk.Push(i);
        }

        for (int i = 0; i < n; ++i)
        {
            int area = heights[i] * (right[i] - left[i]);
            if (area > maxRect.Area)
            {
                maxRect = new Rectangle(area, row - heights[i] + 1, left[i], row, right[i] - 1);
            }
        }

        return maxRect;
    }

    public static Rectangle FindLargestRectangleArea(List<List<int>> matrix)
    {
        int n = matrix.Count;
        if (n == 0) return new Rectangle(0, 0, 0, 0, 0);
        int m = matrix[0].Count;
        if (m == 0) return new Rectangle(0, 0, 0, 0, 0);

        List<int> heights = new List<int>(new int[m]);
        Rectangle maxRect = new Rectangle(0, 0, 0, 0, 0);

        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                heights[j] = matrix[i][j] == 1 ? heights[j] + 1 : 0;
            }
            Rectangle rect = LargestRectangleArea(heights, i);
            if (rect.Area > maxRect.Area)
            {
                maxRect = rect;
            }
        }

        return maxRect;
    }

    public static void Main(string[] args)
    {
        string[] firstLine = Console.ReadLine().Split(' ');
        int n = int.Parse(firstLine[0]);
        int m = int.Parse(firstLine[1]);

        List<List<int>> matrix = new List<List<int>>();
        for (int i = 0; i < n; i++)
        {
            List<int> row = new List<int>();
            string[] line = Console.ReadLine().Split(' ');
            for (int j = 0; j < m; j++)
            {
                row.Add(int.Parse(line[j]));
            }
            matrix.Add(row);
        }

        Rectangle maxRect = FindLargestRectangleArea(matrix);
        Console.WriteLine(maxRect.Area);
        Console.WriteLine($"({maxRect.Top}, {maxRect.Left})");
        Console.WriteLine($"({maxRect.Bottom}, {maxRect.Right})");
    }
}

File: 程式設計：資料結構答案檔/Python/205.py
def largest_rectangle_area(heights):
    stack = []
    max_area = 0
    max_rect = (0, 0, 0, 0)
    i = 0

    while i < len(heights):
        if not stack or heights[i] >= heights[stack[-1]]:
            stack.append(i)
            i += 1
        else:
            while stack and heights[i] < heights[stack[-1]]:
                top = stack.pop()
                width = i if not stack else i - stack[-1] - 1
                area = heights[top] * width
                if area > max_area:
                    max_area = area
                    left = stack[-1] + 1 if stack else 0
                    right = i - 1
                    max_rect = (top, left, top + heights[top] - 1, right)
            stack.append(i)
            i += 1

    while stack:
        top = stack.pop()
        width = i if not stack else i - stack[-1] - 1
        area = heights[top] * width
        if area > max_area:
            max_area = area
            left = stack[-1] + 1 if stack else 0
            right = i - 1
            max_rect = (top, left, top + heights[top] - 1, right)

    return max_area, max_rect

def largest_white_rectangle(matrix):
    if not matrix or not matrix[0]:
        return 0, None, None

    rows = len(matrix)
    cols = len(matrix[0])
    heights = [0] * cols
    max_area = 0
    max_rect = None

    for i in range(rows):
        for j in range(cols):
            if matrix[i][j] == '1':
                heights[j] += 1
            else:
                heights[j] = 0

        area, rect = largest_rectangle_area(heights)
        if area > max_area:
            max_area = area
            top, left, bottom, right = rect
            max_rect = (i - bottom + top, left, i, right)

    return max_area, max_rect

rows,cols = map(int, input().split())
matrix = []
for _ in range(rows):
    row = input().split()
    matrix.append(row)

area, max_rect = largest_white_rectangle(matrix)
print(area)
if area > 0 and max_rect:
    top_left = (max_rect[0], max_rect[1])
    bottom_right = (max_rect[2], max_rect[3])
    print(top_left)
    print(bottom_right)
